File: android/app/src/main/java/org/openbot/env/WiFiDirectConnection.java
package org.openbot.env;

import android.content.Context;
import android.content.IntentFilter;
import android.media.ToneGenerator;
import android.net.wifi.p2p.WifiP2pManager;
import android.util.Log;

import com.abemart.wroup.client.WroupClient;
import com.abemart.wroup.common.WiFiDirectBroadcastReceiver;
import com.abemart.wroup.common.WiFiP2PError;
import com.abemart.wroup.common.WiFiP2PInstance;
import com.abemart.wroup.common.WroupDevice;
import com.abemart.wroup.common.WroupServiceDevice;
import com.abemart.wroup.common.listeners.ClientConnectedListener;
import com.abemart.wroup.common.listeners.DataReceivedListener;
import com.abemart.wroup.common.listeners.ServiceConnectedListener;
import com.abemart.wroup.common.listeners.ServiceDisconnectedListener;
import com.abemart.wroup.common.listeners.ServiceDiscoveredListener;
import com.abemart.wroup.common.messages.MessageWrapper;
import com.abemart.wroup.service.WroupService;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

public class WiFiDirectConnection implements ILocalConnection {
    private static final String TAG = "WiFiDirectConnection";

    private IntentFilter intentFilter = new IntentFilter();
    private WroupClient wroupClient;
    private final String SERVICE_ID = "OPENBOT_SERVICE_ID";
    private boolean connected = false;

    private IDataReceived dataReceivedCallback = null;
    private WiFiDirectBroadcastReceiver wiFiDirectBroadcastReceiver = null;

    @Override
    public void init(Context context) {
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);

        wiFiDirectBroadcastReceiver = WiFiP2PInstance.getInstance(context).getBroadcastReceiver();
        context.registerReceiver(wiFiDirectBroadcastReceiver, intentFilter);
    }

    private void beep() {
        final ToneGenerator tg = new ToneGenerator(6, 100);
        tg.startTone(ToneGenerator.TONE_PROP_BEEP);
    }

    @Override
    public void setDataCallback(IDataReceived dataCallback) {
        dataReceivedCallback = dataCallback;
    }

    @Override
    public void connect(Context context) {
        searchAvailableGroups(context);
    }

    @Override
    public void disconnect(Context context) {
        if (wroupClient != null) {
            wroupClient.disconnect();
        }
        wroupClient = null;
    }

    @Override
    public boolean isConnected() {
        return connected;
    }

    @Override
    public void sendMessage(String message) {
        if (wroupClient != null) {
            try {
                MessageWrapper messageWrapper = new MessageWrapper();
                messageWrapper.setMessage(message);
                messageWrapper.setMessageType(MessageWrapper.MessageType.NORMAL);
                wroupClient.sendMessageToAllClients(messageWrapper);
            } catch (Throwable t) {
                Log.d(TAG, "Something went wrong while trying to send..." + t);
            }
        }
    }

    class DataReceiver implements DataReceivedListener {
        @Override
        public void onDataReceived(MessageWrapper messageWrapper) {
            dataReceivedCallback.dataReceived(messageWrapper.getMessage());
        }
    }

    class ConnectionListener implements ClientConnectedListener {

        @Override
        public void onClientConnected(WroupDevice wroupDevice) {
            connected = true;
            try {
                ControllerToBotEventBus.emitEvent(new JSONObject("{command: \"CONNECTED\"}"));
            } catch (JSONException e) {
                e.printStackTrace();
            }
            beep();
        }
    }

    class ServiceDisconnectionListener implements ServiceDisconnectedListener {

        @Override
        public void onServerDisconnectedListener() {
            connected = false;
            try {
                ControllerToBotEventBus.emitEvent(new JSONObject("{command: \"DISCONNECTED\"}"));
            } catch (JSONException jsonException) {
                jsonException.printStackTrace();
            }
            beep();
        }
    }

    private void searchAvailableGroups(Context context) {

        wroupClient = WroupClient.getInstance(context.getApplicationContext());
        wroupClient.setDataReceivedListener(new DataReceiver());
        wroupClient.setClientConnectedListener(new ConnectionListener());
        wroupClient.setServerDisconnetedListener(new ServiceDisconnectionListener());

        class DiscoveryListener implements ServiceDiscoveredListener {

            @Override
            public void onNewServiceDeviceDiscovered(WroupServiceDevice serviceDevice) {
                Log.i(TAG, "New group found:");
                Log.i(TAG, "\tName: " + serviceDevice.getTxtRecordMap().get(WroupService.SERVICE_GROUP_NAME));

                class ServiceListener implements ServiceConnectedListener {
                    @Override
                    public void onServiceConnected(WroupDevice serviceDevice) {
                        connected = true;
                        try {
                            ControllerToBotEventBus.emitEvent(new JSONObject("{command: \"DISCONNECTED\"}"));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                    }
                }

                if (SERVICE_ID.equals(serviceDevice.getTxtRecordMap().get("GROUP_NAME"))) {
                }

            }

            @Override
            public void onFinishServiceDeviceDiscovered(List<WroupServiceDevice> serviceDevices) {
                Log.i(TAG, "Found '" + serviceDevices.size() + "' groups");
                if (serviceDevices.isEmpty()) {
                    Log.i(TAG, "No groups found");

                    // maybe the controller is not up yet, keep trying
                    searchAvailableGroups(context);
                } else {
                    Log.i(TAG, "Finished discovery");

                    for (WroupServiceDevice serviceDevice : serviceDevices) {
                        String group = serviceDevice.getTxtRecordMap().get("GROUP_NAME");
                        if (SERVICE_ID.equals(group)) {

                            class ServiceConListener implements ServiceConnectedListener {
                                @Override
                                public void onServiceConnected(WroupDevice serviceDevice) {
                                    try {
                                        ControllerToBotEventBus.emitEvent(new JSONObject("{command: \"CONNECTED\"}"));
                                    } catch (JSONException e) {
                                        e.printStackTrace();
                                    }

                                    connected = true;
                                }
                            }
                            if (wroupClient != null) {
                                wroupClient.connectToService(serviceDevice, new ServiceConListener());
                            }
                        }
                    }
                }
            }

            @Override
            public void onError(WiFiP2PError wiFiP2PError) {
                Log.i(TAG, "onError: " + wiFiP2PError.toString());
            }
        }
        if (wroupClient != null) {
            wroupClient.discoverServices(5 * 1000L, new DiscoveryListener());
        }
    }
}

File: android/controller/src/main/java/org/openbot/controller/WiFiDirectConnection.kt
package org.openbot.controller

import android.content.Context
import android.content.IntentFilter
import android.media.ToneGenerator
import android.net.wifi.p2p.WifiP2pManager
import android.util.Log
import com.abemart.wroup.common.WiFiDirectBroadcastReceiver
import com.abemart.wroup.common.WiFiP2PError
import com.abemart.wroup.common.WiFiP2PInstance
import com.abemart.wroup.common.WroupDevice
import com.abemart.wroup.common.listeners.ClientConnectedListener
import com.abemart.wroup.common.listeners.ClientDisconnectedListener
import com.abemart.wroup.common.listeners.DataReceivedListener
import com.abemart.wroup.common.listeners.ServiceRegisteredListener
import com.abemart.wroup.common.messages.MessageWrapper
import com.abemart.wroup.service.WroupService
import org.json.JSONException
import org.openbot.controller.utils.EventProcessor

object WiFiDirectConnection : ILocalConnection {
    var wiFiDirectBroadcastReceiver: WiFiDirectBroadcastReceiver? = null
    var intentFilter = IntentFilter()
    var wroupService: WroupService? = null
    var connected = false
        private set
    private var dataReceivedCallback: IDataReceived? = null
    private const val TAG = "WiFiDirectConnection"
    private const val SERVICE_ID = "OPENBOT_SERVICE_ID"


    override fun init(context: Context) {
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION)
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION)
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION)
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION)
        wiFiDirectBroadcastReceiver = WiFiP2PInstance.getInstance(context).broadcastReceiver
        context.registerReceiver(wiFiDirectBroadcastReceiver, intentFilter)

        wroupService = WroupService.getInstance(context)
        wroupService?.setDataReceivedListener(DataReceiver())
        wroupService?.setClientDisconnectedListener(DisconnectionListener())
        wroupService?.setClientConnectedListener(ConnectionListener())
    }

    override fun setDataCallback(dataCallback: IDataReceived?) {
        dataReceivedCallback = dataCallback
    }

    internal class DataReceiver : DataReceivedListener {
        override fun onDataReceived(messageWrapper: MessageWrapper) {
            dataReceivedCallback?.dataReceived(messageWrapper.message)
        }
    }

    internal class ConnectionListener : ClientConnectedListener {
        override fun onClientConnected(wroupDevice: WroupDevice) {
            Log.i(TAG, "onConnectionResult: connection successful")
            try {
                connected = true

                val event: EventProcessor.ProgressEvents =
                        EventProcessor.ProgressEvents.ConnectionSuccessful
                EventProcessor.onNext(event)

                beep()
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    internal class DisconnectionListener : ClientDisconnectedListener {
        override fun onClientDisconnected(wroupDevice: WroupDevice) {
            Log.i(TAG, "onDisconnected: disconnected...")
            try {
                connected = false

                val event: EventProcessor.ProgressEvents =
                        EventProcessor.ProgressEvents.Disconnected
                EventProcessor.onNext(event)

                beep()
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    override fun connect(context: Context) {
        wroupService!!.registerService(SERVICE_ID, object : ServiceRegisteredListener {
            override fun onSuccessServiceRegistered() {
                Log.i(TAG, "onSuccessServiceRegistered")
            }

            override fun onErrorServiceRegistered(wiFiP2PError: WiFiP2PError) {
                Log.i(TAG, "onErrorServiceRegistered")
            }
        })
    }

    override fun disconnect(context: Context?) {
        if (wroupService != null) {
            wroupService!!.disconnect()
        }
    }

    override fun isConnected(): Boolean {
        return connected
    }

    override fun sendMessage(message: String?) {
        val messageWrapper = MessageWrapper()
        messageWrapper.message = message
        messageWrapper.messageType = MessageWrapper.MessageType.NORMAL
        wroupService!!.sendMessageToAllClients(messageWrapper)
    }

    override fun start() {
        TODO("Not yet implemented")
    }

    override fun stop() {
        TODO("Not yet implemented")
    }

    private fun beep() {
        val tg = ToneGenerator(6, 100)
        tg.startTone(ToneGenerator.TONE_PROP_BEEP)
    }
}
